src/hotspot/share/ci/ciMethod.cpp

```diff
@@ -682,12 +682,19 @@ ciMethod* ciMethod::find_monomorphic_target(ciInstanceKlass* caller,
     return nullptr;
   }
 
+  VM_ENTRY_MARK;
+
   ciMethod* root_m = resolve_invoke(caller, actual_recv, check_access, true /* allow_abstract */);
   if (root_m == nullptr) {
     // Something went wrong looking up the actual receiver method.
     return nullptr;
   }
 
+  // Redefinition support.
+  if (this->is_old() || root_m->is_old()) {
+    return nullptr;
```

Comment:
IMO you can safely drop this particular check. The one after `Dependencies::find_unique_concrete_method()` should be enough to preserve the invariant (`target == cha_monomorphic_target`) .

But thinking more about it, now I'm curious what happens when an old method is actually encountered. The fix conservatively rejects possible inlining opportunity, but it seems it doesn't invalidate resulting nmethod anymore. So, no recompilation attempt follows to recuperate that.

We could either record a evol dependency on a stale `Method` (to fail during nmethod installation step) or fail-fast the compilation (probably, implies additional checks to propagate the failure status).

Reply:
Even if we check for stale Methods in various places, including invoke(), there is nothing to prevent the method from going stale after the last spot-check.  My understanding was that we already handle stale metadata as a precondition to creating the nmethod.  If we have a loophole there that lets stale metadata get through, then that's a separate existing bug for C1 and C2.
I was tempted to add a bailout, but the reason would be as a performance improvement to short-circuit wasted work, not to correct a stale metadata problem.

Reply:
> IMO you can safely drop this particular check. The one after Dependencies::find_unique_concrete_method() should be enough to preserve the invariant (target == cha_monomorphic_target) .

If I do that, then I can also revert the VM state changes.  However, I wasn't able to convince myself that this check is not needed.  If we end up returning root_m here as cha_monomorphic_target, it seems possible that it could be a new version of the method, and then target == cha_monomorphic_target would fail.

Reply:
C1 does call 
```
2159            dependency_recorder()->assert_evol_method(inline_target);
```
which in the CHA case would be `cha_monomorphic_target`, not `target`, so it looks like we may not detect if `target` is stale as long as `cha_monomorphic_target` is not.  It seems like a minor loophole, but I'm not sure what kind of problems it could cause, especially if the bytecodes of `target` are not used.

Reply:
The scenario which concerns me is performance-related. If CHA conservatively disables inlining when concurrent class redefinition takes place during parsing, then there are no mechanism in place to recuperate possible loss of performance. Normally, if inlined method is redefined later during compilation, nmethod installation fails during dependency validation. But here no inlining happens (missed optimization opportunity) and call sites in generated code are resolved based on symbolic information (except rare cases when resolved method is attached to the call site, see `SharedRuntime::find_callee_info_helper()` for details), so there are no guarantees the stale `Method*` is recorded.

I agree that the window for such sequence of events is narrow, but it may be a source of surprising performance anomalies in rare cases.

Reply:
OK, it sounds like we have two choices.  Either record an evol dependency every time we short-circuit an optimization based on is_old, or bail out.  I vote for bailing out.

Reply:
I vote for bail out.

Reply:
I prefer bailing out as well, but, please, check it doesn't mark the root method as non-compilable.